File: sdk/typescript/src/api/directory.ts
import type { HttpClient } from "../http.js";
import type {
  AgentCard,
  AgentSearchResponse,
  AgentQueryParams,
  DirectoryIdentityListingsResponse,
  ExtendedAgentCard,
  IdentityListingQueryParams,
  ResolveResponse,
  ReverseResponse,
} from "../types/index.js";
import {
  validateAgentCard,
  validateAgentQueryParams,
  validateExtendedAgentCard,
  validateIdentityListingQueryParams,
} from "../validation.js";

export class DirectoryApi {
  constructor(private readonly http: HttpClient) {}

  listAgents(params?: AgentQueryParams): Promise<{ agents: Array<AgentCard> }> {
    validateAgentQueryParams(params);
    return this.http.get<{ agents: Array<AgentCard> }>(
      "/directory/agents",
      params as Record<string, unknown>,
    );
  }

  getAgent(agentId: string): Promise<AgentCard> {
    return this.http.get<AgentCard>(
      `/directory/agents/${encodeURIComponent(agentId)}`,
    );
  }

  getExtendedAgent(agentId: string): Promise<ExtendedAgentCard> {
    return this.http.getDirectoryAuth<ExtendedAgentCard>(
      `/directory/agents/${encodeURIComponent(agentId)}/extended`,
    );
  }

  upsertExtendedAgent(
    agentId: string,
    card: ExtendedAgentCard,
  ): Promise<ExtendedAgentCard> {
    validateExtendedAgentCard({ ...card, agentId });
    return this.http.putDirectoryAuth<ExtendedAgentCard>(
      `/directory/agents/${encodeURIComponent(agentId)}/extended`,
      card,
    );
  }

  upsertAgent(agentId: string, card: AgentCard): Promise<AgentCard> {
    validateAgentCard({ ...card, agentId, cryptoId: card.cryptoId || agentId });
    return this.http.putDirectoryAuth<AgentCard>(
      `/directory/agents/${encodeURIComponent(agentId)}`,
      card,
    );
  }

  deleteAgent(agentId: string): Promise<void> {
    return this.http.deleteDirectoryAuth<void>(
      `/directory/agents/${encodeURIComponent(agentId)}`,
    );
  }

  listIdentities(
    params?: IdentityListingQueryParams,
  ): Promise<DirectoryIdentityListingsResponse> {
    validateIdentityListingQueryParams(params);
    return this.http.get<DirectoryIdentityListingsResponse>(
      "/directory/identities",
      params as Record<string, unknown>,
    );
  }

  resolve(name: string): Promise<ResolveResponse> {
    return this.http.get<ResolveResponse>(
      `/directory/resolve/${encodeURIComponent(name)}`,
    );
  }

  reverse(cryptoId: string): Promise<ReverseResponse> {
    return this.http.get<ReverseResponse>(
      `/directory/reverse/${encodeURIComponent(cryptoId)}`,
    );
  }

  skills(params?: {
    q?: string;
    limit?: number;
    cursor?: string;
  }): Promise<AgentSearchResponse> {
    validateAgentQueryParams(params);
    return this.http.get<AgentSearchResponse>(
      "/directory/skills",
      params as Record<string, unknown>,
    );
  }
}

File: sdk/rust/src/api/directory.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::http::{HttpClient, to_body, to_query};
use crate::types::{AgentCard, AgentQueryParams, ExtendedAgentCard, ResolveResponse, ReverseResponse};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentListResponse {
    pub agents: Vec<AgentCard>,
}

pub struct DirectoryApi {
    http: Arc<HttpClient>,
}

impl DirectoryApi {
    pub fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn list_agents(
        &self,
        params: Option<&AgentQueryParams>,
    ) -> Result<AgentListResponse> {
        let query = params.map(to_query).transpose()?;
        self.http.get("/directory/agents", query.as_ref()).await
    }

    pub async fn get_agent(&self, agent_id: &str) -> Result<AgentCard> {
        let path = format!("/directory/agents/{}", urlencoding::encode(agent_id));
        self.http.get(&path, None).await
    }

    pub async fn get_extended_agent(&self, agent_id: &str) -> Result<ExtendedAgentCard> {
        let path = format!(
            "/directory/agents/{}/extended",
            urlencoding::encode(agent_id)
        );
        self.http.get_auth(&path, None).await
    }

    pub async fn upsert_agent(&self, agent_id: &str, card: &AgentCard) -> Result<AgentCard> {
        let path = format!("/directory/agents/{}", urlencoding::encode(agent_id));
        self.http.put(&path, Some(&to_body(card)?)).await
    }

    pub async fn delete_agent(&self, agent_id: &str) -> Result<()> {
        let path = format!("/directory/agents/{}", urlencoding::encode(agent_id));
        self.http.delete(&path, None).await
    }

    pub async fn resolve(&self, name: &str) -> Result<ResolveResponse> {
        let path = format!("/directory/resolve/{}", urlencoding::encode(name));
        self.http.get(&path, None).await
    }

    pub async fn reverse(&self, crypto_id: &str) -> Result<ReverseResponse> {
        let path = format!("/directory/reverse/{}", urlencoding::encode(crypto_id));
        self.http.get(&path, None).await
    }
}
